core/services/ocr2/plugins/rebalancer/plugin_test.go

```diff
@@ -2,67 +2,1221 @@ package rebalancer
+	rebalancermocks "github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/rebalancer/rebalancermocks"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/rebalancer/testhelpers"
 )
 
-type mockDeps struct {
-	mockFactory           *rebalancer_mocks.Factory
-	mockRebalancer        *rebalancer_mocks.Rebalancer
-	mockDiscovererFactory *discoverermocks.Factory
-	mockBridgeFactory     *bridgemocks.Factory
+func TestPlugin_Query(t *testing.T) {
+	t.Run("query should always be empty", func(t *testing.T) {
+		p := newPluginWithMocksAndDefaults(t)
+		q, err := p.plugin.Query(context.Background(), ocr3types.OutcomeContext{})
+		assert.NoError(t, err)
+		assert.Empty(t, q)
+
+		q, err = p.plugin.Query(context.Background(), ocr3types.OutcomeContext{SeqNr: 1234})
+		assert.NoError(t, err)
+		assert.Empty(t, q)
+	})
+}
+
+func TestPlugin_Observation(t *testing.T) {
+	ctx := testutils.Context(t)
+
+	testCases := []struct {
+		name            string
+		seqNr           uint64
+		observedGraph   func(t *testing.T) (graph.Graph, error)
+		previousOutcome models.Outcome
+		bridges         map[[2]models.NetworkSelector]func(t *testing.T) (bridge.Bridge, error)
+		expObservation  models.Observation
+		expErr          func(t *testing.T, err error)
+	}{
+		{
+			name:  "no neighboring networks",
+			seqNr: 1,
+			observedGraph: func(t *testing.T) (graph.Graph, error) {
+				return graph.NewGraph(), nil
+			},
+			previousOutcome: models.Outcome{},
+			bridges:         nil,
+			expObservation: models.NewObservation(
+				[]models.NetworkLiquidity{},
+				[]models.Transfer{},
+				[]models.PendingTransfer{},
+				[]models.Edge{},
+				[]models.ConfigDigestWithMeta{},
+			),
+			expErr: nil,
+		},
+		{
+			name:  "two networks that generate a full report",
+			seqNr: 1,
+			observedGraph: func(t *testing.T) (graph.Graph, error) {
+				g := graph.NewGraph()
+				g.AddNetwork(networkA, graph.Data{
+					Liquidity:         big.NewInt(1000),
+					TokenAddress:      tokenX,
+					RebalancerAddress: rebalancerA,
+					NetworkSelector:   networkA,
+					ConfigDigest:      cfgDigest1,
+				})
+				g.AddNetwork(networkB, graph.Data{
+					Liquidity:         big.NewInt(2000),
+					TokenAddress:      tokenY,
+					RebalancerAddress: rebalancerB,
+					NetworkSelector:   networkB,
+					ConfigDigest:      cfgDigest2,
+				})
+				assert.NoError(t, g.AddConnection(networkA, networkB))
+				assert.NoError(t, g.AddConnection(networkB, networkA))
+				return g, nil
+			},
+			previousOutcome: models.Outcome{
+				ProposedTransfers: []models.ProposedTransfer{
+					{
+						From:   networkB,
+						To:     networkA,
+						Amount: ubig.New(big.NewInt(123)),
+					},
+				},
+			},
+			bridges: map[[2]models.NetworkSelector]func(t *testing.T) (bridge.Bridge, error){
+				[2]models.NetworkSelector{networkA, networkB}: func(t *testing.T) (bridge.Bridge, error) {
```

Comment:
Don't need to specify the type in this instance

```suggestion
				{networkA, networkB}: func(t *testing.T) (bridge.Bridge, error) {
```